### PR TITLE
Removing mutable function arguments

### LINKED ISSUE #20068: Use flake8-bugbear on Travis and fix warnings
See e.g. https://github.com/sympy/sympy/pull/20061#discussion_r485996905

Currently flake8 is used on most files under the sympy folder. Mostly it warns about unused variables and imports in code and also non-existent variables which can help to pick up on typos and helps to keep the imports up to date. This was recently extended to docstrings (#19374, #19406).

It would be good to pick up on some other things though such as the use of mutable default arguments. There is flake-mutable but it seems that it basically doesn't work e.g. https://github.com/ebeweber/flake8-mutable/issues/21. The recommendation there is to use flake8-bugbear instead.

I've listed all the warnings that show up when using flake8-bugbear below. If we agree that they should all be fixed then I think that maybe this can be marked as easy to fix although it's not as simple as #19374. Once the warnings are fixed we could enable the use of flake8-bugbear on Travis with:
```diff
diff --git a/.travis.yml b/.travis.yml
index b21eef3c3a..a578f95460 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -214,7 +214,7 @@ before_install:
       coverage debug sys;
     fi
   - if [[ "${TEST_FLAKE8}" == "true" ]]; then
-      pip install flake8;
+      pip install flake8 flake8-bugbear;
     fi
   - if [ "$TRAVIS_PYTHON_VERSION" == "3.8" ]; then
         pip uninstall -y Cython;
```

We can test this with
```console
$ pip install flake8-bugbear
```
The warnings from flake8-bugbear need to be enabled in setup.cfg with this diff:
```diff
diff --git a/setup.cfg b/setup.cfg
index 112c744bb4..f86e990e45 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -5,7 +5,7 @@ long_description_content_type = text/markdown
 [flake8]
 doctests = True
 ignore = F403
-select = F,E722
+select = F,E722,B
 exclude = sympy/parsing/latex/_antlr/*,
           sympy/parsing/autolev/_antlr/*,
           sympy/parsing/autolev/test-examples/*,
```
Then you can run
```console
$ flake8 sympy
```
and you will see a list of warnings. We can count the warning types with:
```
$ flake8 sympy > bugbear.txt
```
And then this Python script:
```python
from itertools import chain
import re
from collections import Counter
from pprint import pprint

with open('bugbear.txt') as fin:
    words = chain.from_iterable(map(str.split, fin))
    codes = (w for w in words if re.match('B\d\d\d', w))
    counts = Counter(codes)

pprint(counts)
```
There are 493 total warnings and the counts for individual warning codes are
```
Counter({'B007': 300,
         'B006': 76,
         'B305': 37,
         'B011': 29,
         'B008': 28,
         'B009': 7,
         'B004': 5,
         'B301': 5,
         'B010': 4,
         'B012': 1,
         'B014': 1})
```

B007 - unused loop variable
-----

There are 300 warnings of this type. An example:
```
sympy/series/gruntz.py:464:14: B007 Loop control variable 'expr' not used within the loop body. If this is intended, start the name with an underscore.
```
This is suggesting that if the loop variable is not used then it should have the throwaway name `_` or otherwise should begin with an underscore. The obvious fix is this:
```diff
diff --git a/sympy/series/gruntz.py b/sympy/series/gruntz.py
index 712e073f1d..487230990c 100644
--- a/sympy/series/gruntz.py
+++ b/sympy/series/gruntz.py
@@ -461,7 +461,7 @@ def moveup2(s, x):
     r = SubsSet()
     for expr, var in s.items():
         r[expr.xreplace({x: exp(x)})] = var
-    for var, expr in s.rewrites.items():
+    for var, _ in s.rewrites.items():
         r.rewrites[var] = s.rewrites[var].xreplace({x: exp(x)})
     return r
```
However probably a better fix in this particular case is just not using `.items()`:
```diff
diff --git a/sympy/series/gruntz.py b/sympy/series/gruntz.py
index 712e073f1d..d27230acfd 100644
--- a/sympy/series/gruntz.py
+++ b/sympy/series/gruntz.py
@@ -461,7 +461,7 @@ def moveup2(s, x):
     r = SubsSet()
     for expr, var in s.items():
         r[expr.xreplace({x: exp(x)})] = var
-    for var, expr in s.rewrites.items():
+    for var in s.rewrites:
         r.rewrites[var] = s.rewrites[var].xreplace({x: exp(x)})
     return r
```

B006 - mutable default argument
----

There are 76 warnings of this type. Example:
```
sympy/core/add.py:417:39: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
```
The fix could be this:
```diff
diff --git a/sympy/core/add.py b/sympy/core/add.py
index a37140b836..3676967abc 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -414,7 +414,9 @@ def _matches_simple(self, expr, repl_dict):
             return terms[0].matches(expr - coeff, repl_dict)
         return
 
-    def matches(self, expr, repl_dict={}, old=False):
+    def matches(self, expr, repl_dict=None, old=False):
+        if repl_dict is None:
+            repl_dict = {}
         return self._matches_commutative(expr, repl_dict, old)
```
Alternatively `matches` could pass the `None` on to `_matches_commutative` which could check for `None` in a similar way. Using mutable defaults like this is problematic and can lead to bugs if `repl_dict` is ever modified inside `_matches_commutative` or is ever returned directly to a caller from any function. We should definitely fix all of these.

B305 - Use `next(obj)` rather than `obj.next()`
----------

This is warning about using the Python 2 `.next()` iterator method rather than the `next` function in Python 3. Example:
```
sympy/printing/pretty/stringpict.py:109:16: B305 `.next()` is not a thing on Python 3. Use the `next()` builtin. For Python 2 compatibility, use `six.next()`.
```
There are a bunch of warnings about this in stringpict. They seem to be false positives in the sense that these classes have a `next` method that is not intended to be used for iteration. I guess that having a method name that conflicts with a standard method name like `next` is not ideal but then again `next` has been removed in Python 3 in favour of `__next__` precisely because it's not good to have special method names that can easily conflict with normal method names. We could select to ignore this error code.

B011 - Don't assert False
------------

This is warning about using `assert False` and suggesting to use `raise AssertionError` instead. Example:
```
sympy/utilities/tests/test_misc.py:98:9: B011 Do not call assert False since python -O removes these calls. Instead callers should raise AssertionError().
```
The example is from test code that looks like this:
```python
def test_translate_args():
    try:
        translate(None, None, None, 'not_none')
    except ValueError:
        pass # Exception raised successfully
    else:
        assert False
```
That could be changed to `raise AssertionError()` but again a better fix is possible. Whoever wrote that must have not known that we have the `raises` function for precisely this purpose. The fix is:
```diff
diff --git a/sympy/utilities/tests/test_misc.py b/sympy/utilities/tests/test_misc.py
index 5d57fad968..6eb747ae12 100644
--- a/sympy/utilities/tests/test_misc.py
+++ b/sympy/utilities/tests/test_misc.py
@@ -4,6 +4,7 @@
 import os
 
 from sympy.utilities.misc import translate, replace, ordinal, rawlines, strlines
+from sympy.testing.pytest import raises
 
 def test_translate():
     abc = 'abc'
@@ -90,12 +91,7 @@ def test_strlines():
 
 
 def test_translate_args():
-    try:
-        translate(None, None, None, 'not_none')
-    except ValueError:
-        pass # Exception raised successfully
-    else:
-        assert False
+    raises(ValueError, lambda: translate(None, None, None, 'not_none'))
 
     assert translate('s', None, None, None) == 's'
```

B008 - No function calls for default argument values
-----------

Example:
```
sympy/testing/runtests.py:1003:21: B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time function call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
```
This is complaining about
```python
def sympytestfile(filename, module_relative=True, name=None, package=None,
             globs=None, verbose=None, report=True, optionflags=0,
             extraglobs=None, raise_on_error=False,
             parser=pdoctest.DocTestParser(), encoding=None):
```
The suggestion is to do
```python
default_parser = pdoctest.DocTestParser()

def sympytestfile(filename, module_relative=True, name=None, package=None,
             globs=None, verbose=None, report=True, optionflags=0,
             extraglobs=None, raise_on_error=False,
             parser=default_parser, encoding=None):
```
That seems a reasonable change to me because it's slightly more explicit but I don't have strong views. Probably this is something that trips people up sometimes so maybe it would be good to have a test for it.

B009, B010 - Don't use getattr/serattr with constant attribute name
-------------

Example:
```
sympy/external/tests/test_sage.py:71:24: B009 Do not call getattr with a constant attribute value, it is not any safer than normal property access.
```
This saying that e.g. `getattr(x, 'value')` should just be `x.value`. I agree with that.

B004 - Don't use hasattr for `__call__`
------------

Example:
```
sympy/core/expr.py:3967:16: B004 Using `hasattr(x, '__call__')` to test if `x` is callable is unreliable. If `x` implements custom `__getattr__` or its `__call__` is itself not callable, you might get misleading results. Use `callable(x)` for consistent results.
```
There is a Python 3 builtin for checking if an object is callable which we can use now that Python 2 is dropped:
https://docs.python.org/3/library/functions.html#callable

B301 - don't use iteritems etc
----------

Example:
```
sympy/testing/tests/diagnose_imports.py:161:27: B301 Python 3 does not include `.iter*` methods on dictionaries. Remove the `iter` prefix from the method name. For Python 2 compatibility, prefer the Python 3 equivalent unless you expect the size of the container to be large or unbounded. Then use `six.iter*` or `future.utils.iter*`.
```
This is saying to use `d.keys()` rather than `d.iterkeys()` which is only present in Python 2.x. I agree with that since it won't work in Python 3 otherwise. The fact this hasn't already come up suggests that `diagnose_imports.py` isn't used or tested so we could also consider removing it.

B012 - Don't return from finally
--------

Example:
```
sympy/testing/runtests.py:284:9: B012 return/continue/break inside finally blocks cause exceptions to be silenced. Exceptions should be silenced in except blocks. Control statements can be moved outside the finally block.
```
Seems reasonable.

B014 - redundant exception types
--------

Example:
```
sympy/testing/runtests.py:2064:13: B014 Redundant exception types in `except (OSError, IOError):`.  Write `except ():`, which catches exactly the same exceptions.
```
This is reasonable although the error message it garbled. The point is this:
```
>>> issubclass(IOError, OSError)
True
```
That means that `except OSError` will catch any instance of `IOError` so it doesn't also need to be listed.

### PR DESCRIPTION
**Pull request by sayandip18**
#### References to other Issues or PRs

 #20068 

#### Brief description of what is fixed or changed
The aim is to do away with all the mutable default arguments used in functions. Each of the B006 warnings are examined individually to look for possible fixes.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
It looks like somehow spaces have been converted into tabs in `sympy/combinatorics/fp_groups.py`

**Comment by sayandip18** [comment]
I guess I have to run `sympy\testing\tests\test_code_quality.py` locally before pushing any more commits.

File: sympy/core/function.py [hunk 1]
```diff
@@ -1022,7 +1022,9 @@ def __init__(cls, name, **assumptions):
             nargs = FiniteSet(*nargs)
         cls.nargs = nargs
 
-    def matches(self, expr, repl_dict={}, old=False):
+    def matches(self, expr, repl_dict=None, old=False):
+        if repl_dict is None:
+            repl_dict - {}
```
Comment (oscarbenjamin) [inline sympy/core/function.py:1027]: This should be `=`

**Comment by oscarbenjamin** [comment]
Yes, you can run `bin/test quality`.

Right now it looks like a file has trailing whitespace which you can fix with `bin/strip_whitespace`.

File: sympy/core/function.py (hunk 1 above)
Comment (sayandip18) [reply sympy/core/function.py:1027]: Fixing it. My bad!

**Comment by sayandip18** [comment]
```
>>> import subprocess
>>> subprocess.call('C:\Users\SANDIP\sympy\bin\strip_whitespace')
  File "<stdin>", line 1
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 2-3: truncated \UXXXXXXXX escape
```

I'm getting this error. What am I doing wrong?

**Comment by sayandip18** [comment]
I reverted the changes for [`simplify/simplify.py`](https://github.com/sympy/sympy/blob/b6c328e2f057ba81513d19a1451ace358ecc8233/sympy/simplify/simplify.py#L40) and kept it as it is because it was throwing [`AssertionError`](https://travis-ci.org/github/sympy/sympy/jobs/726591608#L509)

**Comment by oscarbenjamin** [comment]
What test was giving `AssertionError`?

**Comment by sayandip18** [comment]
[This](https://github.com/sympy/sympy/blob/b6c328e2f057ba81513d19a1451ace358ecc8233/sympy/simplify/tests/test_simplify.py#L282) in `test_simplify.py`

**Comment by oscarbenjamin** [comment]
I think this should work:
```diff
diff --git a/sympy/simplify/simplify.py b/sympy/simplify/simplify.py
index d122999e3f..481758a26f 100644
--- a/sympy/simplify/simplify.py
+++ b/sympy/simplify/simplify.py
@@ -37,7 +37,7 @@
 
 
 
-def separatevars(expr, symbols=[], dict=False, force=False):
+def separatevars(expr, symbols=(), dict=False, force=False):
     """
     Separates variables in an expression, if possible.  By
     default, it separates with respect to all symbols in an
```

**Comment by oscarbenjamin** [comment]
I think we can use this to enable bugbear only for B006:
```diff
diff --git a/setup.cfg b/setup.cfg
index 112c744bb4..e61daa9ec6 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -5,7 +5,7 @@ long_description_content_type = text/markdown
 [flake8]
 doctests = True
 ignore = F403
-select = F,E722
+select = F,E722,B006
 exclude = sympy/parsing/latex/_antlr/*,
           sympy/parsing/autolev/_antlr/*,
           sympy/parsing/autolev/test-examples/*,
```

**Comment by oscarbenjamin** [comment]
The OP should be edited so it doesn't say that it "fixes" the issue as this is only a partial fix for the issue.

The commits in this PR will need to be squashed because of the tab/space changes in `fp_groups.py` that were reverted. Committing changes and then reverting them again still affects the "git blame" function since most lines in that file will show as having been last modified in the "tabs" commit. When squashing the commits you can give a new commit message so try to come up with something that will make sense to someone who looks at the message in say a few years time e.g.:
```
maint: don't use mutable default arguments

This commit removes all cases of using mutable default function arguments
from the codebase. It also adds to the flake8-bugbear to the flake8 configuration
so that flake8 will pick up on any examples of mutable defaults when the tests
are run in CI.
```

File: sympy/simplify/simplify.py [hunk 2]
```diff
@@ -37,7 +37,7 @@
 
 
 
-def separatevars(expr, symbols=None, dict=False, force=False):
+def separatevars(expr, symbols=[], dict=False, force=False):
```
Comment (jksuom) [inline sympy/simplify/simplify.py]: `symbols=()` would be immutable.

Comment (sayandip18) [reply sympy/simplify/simplify.py]: Thanks! Doing it.

**Comment by sayandip18** [comment]
This has become messy because of some merge conflicts. Should I close this to start a new PR or keep modifying this?

**Comment by oscarbenjamin** [comment]
> Should I close this to start a new PR or keep modifying this?

You don't need to close the PR. I suggest to keep working on this one. When it is complete then you can squash the commits so the merge will be gone.

File: sympy/combinatorics/homomorphisms.py [hunk 3]
```diff
@@ -264,7 +264,7 @@ def invert_subgroup(self, H):
                     P = PermutationGroup(gens)
         return P
 
-def homomorphism(domain, codomain, gens, images=[], check=True):
+def homomorphism(domain, codomain, gens, images=None, check=True):
```
Comment (eric-wieser) [inline sympy/combinatorics/homomorphisms.py:267]:
```suggestion
def homomorphism(domain, codomain, gens, images=(), check=True):
```

File: sympy/combinatorics/homomorphisms.py [hunk 4]
```diff
@@ -280,6 +280,8 @@ def homomorphism(domain, codomain, gens, images=[], check=True):
     define a homomorphism.
 
     '''
+    if images is None:
+        images = []
```
Comment (eric-wieser) [inline sympy/combinatorics/homomorphisms.py:284]:
```suggestion
```

File: sympy/combinatorics/polyhedron.py [hunk 5]
```diff
@@ -379,6 +379,10 @@ def __new__(cls, corners, faces=[], pgroup=[]):
         .. [1] www.ocf.berkeley.edu/~wwu/articles/platonicsolids.pdf
 
         """
+        if pgroup is None:
+            pgroup = []
+        if faces is None:
+            faces = []
```
Comment (eric-wieser) [inline sympy/combinatorics/polyhedron.py:385]:
```suggestion
```

File: sympy/combinatorics/polyhedron.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ class Polyhedron(Basic):
     """
     _edges = None
 
-    def __new__(cls, corners, faces=[], pgroup=[]):
+    def __new__(cls, corners, faces=None, pgroup=None):
```
Comment (eric-wieser) [inline sympy/combinatorics/polyhedron.py:30]:
```suggestion
    def __new__(cls, corners, faces=(), pgroup=()):
```

File: sympy/core/mul.py [hunk 7]
```diff
@@ -1001,13 +1003,15 @@ def _matches_expand_pows(arg_list):
...
+    def _matches_noncomm(nodes, targets, repl_dict=None):
         """Non-commutative multiplication matcher.
 
         `nodes` is a list of symbols within the matcher multiplication
         expression, while `targets` is a list of arguments in the
         multiplication expression being matched against.
         """
+        if repl_dict is None:
+            repl_dict = {}
         repl_dict = repl_dict.copy()
```
Comment (eric-wieser) [inline sympy/core/mul.py:1015]: Note that the change here is completely redudant, as the copy on the next line ensures that no mutation can occur

**Review by eric-wieser — CHANGES_REQUESTED** [review]
I'm not really a fan of this type of change, as I discussed in https://github.com/numba/numba/pull/5811, and just opened at https://github.com/PyCQA/flake8-bugbear/issues/137.

Essentially, one of two things is true:

* These functions do not actually mutate their argument; this change adds complexity and makes the default values harder to work out from the signature.
  At a glance, many of the functions touched by this PR already handle this - anywhere you see `arg = list(arg)` or `arg = arg.copy()`, there is no danger of argument mutation
* These functions do mutate their argument; this change misses the forest for the trees, as callers who pass a non-default value will find that it gets mutated too.

**Comment by oscarbenjamin** [comment]
> * there is no danger of argument mutation

I think that there is always a danger of mutation. Even if the code is correct now it is easy for it to become incorrect later when someone makes a seemingly innocent change. I have seen bugs resulting from such mutation that can be hard to identify e.g. in #19261 it was not easy for experienced contributors to pin down the bug from here:
https://github.com/sympy/sympy/blob/0180df08428d4c9ccff124fd2984720e8409f46f/sympy/core/basic.py#L1556-L1589

The effect of this can be seen in SymPy 1.6:
```julia
In [1]: exp(x).matches(exp(y))                                                                                                                 

In [2]: d = exp(x).matches(exp(x))                                                                                                             

In [3]: d[x] = y                                                                                                                               

In [4]: exp(x).matches(exp(y))                                                                                                                 
Out[4]: {x: y}
```

In some cases there might be a better fix than using `is None` and mostly having an immutable default is a better option when it works.

**Comment by eric-wieser** [comment]
Thanks for that example. I think the point I'm trying to make is that it's not accidentally-mutated defaults that are dangerous, but accidentally-mutated argument values. If we blindly apply a `if x is None: x = {}` strategy, we fix all the problems of the former category while making the latter category even harder to find.

> mostly having an immutable default

This approach is fine by me - it's the `is None` approach that raises the red flags. Replacing `[]` with `()`, `{}` with `MappingProxyType({})`, and `set()` with `frozenset()` solves the mutable default problem, and also raises an exception if anything actually tries to mutate the argument.

File: sympy/combinatorics/fp_groups.py [hunk 8]
```diff
@@ -17,7 +17,9 @@
 
 
 @public
-def fp_group(fr_grp, relators=[]):
+def fp_group(fr_grp, relators=None):
+	if relators is None:
+		relators = []
```
Comment (eric-wieser) [inline sympy/combinatorics/fp_groups.py:22]:
```suggestion
def fp_group(fr_grp, relators=()):
	  relators = list(relators)
```
which avoids mutating the argument when it is explicitly provided too.

**Comment by oscarbenjamin** [comment]
> If we blindly apply a `if x is None: x = {}` strategy, we fix all the problems of the former category while making the latter category even harder to find.

I agree.

File: sympy/combinatorics/fp_groups.py (hunk 8 above)
Comment (sayandip18) [reply sympy/combinatorics/fp_groups.py:22]: This is a previous commit actually. I force pushed to it as there were a lots of issues with my PR including merging conflicts. I'll soon make the changes according to what you suggested.

**Comment by sayandip18** [comment]
The review was requested by mistake. The tests are yet to pass as of now.

File: sympy/simplify/hyperexpand.py [hunk 9]
```diff
@@ -1964,6 +1964,7 @@ def _hyperexpand(func, z, ops0=[], z0=Dummy('z0'), premult=1, prem=0,
     is multiplied by premult. Then ops0 is applied.
     premult must be a*z**prem for some a independent of z.
     """
+    ops0 = list(ops0)
```
Comment (eric-wieser) [inline sympy/simplify/hyperexpand.py:1967]: What happens if you skip this?

Comment (sayandip18) [reply sympy/simplify/hyperexpand.py:1967]: Nothing, it seems. I ran several tests and all passed. I guess then I should delete this line in the next and the final commit before squashing.

File: sympy/core/expr.py [hunk 10]
```diff
@@ -3963,7 +3963,8 @@ def unchanged(func, *args):
 
 
 class ExprBuilder:
-    def __init__(self, op, args=[], validator=None, check=True):
+    def __init__(self, op, args=(), validator=None, check=True):
+        args = list(args)
```
Comment (eric-wieser) [inline sympy/core/expr.py:3967]: Does it work without this line?

Comment (sayandip18) [reply sympy/core/expr.py:3967]: No. `test_expr.py` fails.

```
>>> sympy.test("sympy/core/tests/test_expr.py")
========================================================================= test process starts =========================================================================
executable:         C:\Users\SANDIP\anaconda3\python.exe  (3.7.6-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python
numpy:              1.18.1
random seed:        44792414
hash randomization: on (PYTHONHASHSEED=3963841612)

sympy\core\tests\test_expr.py[100] ................................f.......................................................f.........E.                          [FAIL]

_______________________________________________________________________________________________________________________________________________________________________
___________________________________________________________ sympy\core\tests\test_expr.py:test_ExprBuilder ____________________________________________________________
Traceback (most recent call last):
  File "c:\users\sandip\sympy\sympy\core\tests\test_expr.py", line 2111, in test_ExprBuilder
    eb.args.extend([x, x])
AttributeError: 'tuple' object has no attribute 'extend'

============================================ tests finished: 97 passed, 2 expected to fail, 1 exceptions, in 14.16 seconds ============================================
DO *NOT* COMMIT!
False
```

**Comment by sayandip18** [comment]
Squashed into one commit. I hope this is ready to be merged.

**Comment by czgdp1807** [comment]
@sayandip18 Would you like to complete this PR? If yes then, please resolve the conflicts. Please do not close this PR. Thanks for your contributions.

**Comment by sayandip18** [comment]
Yeah sure I would. Actually I had totally forgotten about this PR. I'll resolve the conflicts and squash as soon as I fix the issue at hand right now.

**Pull request closed without merging**